store/tikv/2pc.go

```diff
@@ -214,14 +230,16 @@ func (c *twoPhaseCommitter) doActionOnKeys(bo *Backoffer, action twoPhaseCommitA
 		sizeFunc = c.keyValueSize
 	}
 	// Make sure the group that contains primary key goes first.
-	batches = appendBatchBySize(batches, firstRegion, groups[firstRegion], sizeFunc, txnCommitBatchSize)
-	delete(groups, firstRegion)
+	if len(groups[primaryRegion]) > 0 {
+		batches = appendBatchBySize(batches, primaryRegion, groups[primaryRegion], sizeFunc, txnCommitBatchSize)
+		batches[0].primary = true
```

Comment:
`primaryKey` may not in the first batch.

Reply:
No, it may be not.
There two cases to handle in these lines of code.
1. The first time to commit a batch. In this case, the `GroupKeysByRegion` returns the primary region because that the `keys` contains the `primary key`, so the first batch contains the `primary key`.

2. A `re-split` routine is executed. This case is caused when submitting a secondary batch in the background with a failure caused by the `region error`. In this case, the `keys` to be grouped to batches does not contain the `primary key`, and there is not a `primaryRegion`. 

All these above conditions are checked by `len(groups[primaryRegion]) > 0`.  In fact, I am still wondering if this is a safety check. If a region id of 0 is valid, the check may be false positive.

Reply:
If there too many keys in the same region with the primary key, `appendBatchBySize` will split them into multiple batches. After split, the primary key may not in the first batch. So it is incorrect to mark the first batch as primary.